destination/destination.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright © 2022 Meroxa, Inc. & Gophers Lab Technologies Pvt. Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package destination
+
+import (
+	"context"
+	"sync"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/conduitio/conduit-connector-zendesk/zendesk"
+)
+
+type Writer interface {
+	Write(ctx context.Context, records []sdk.Record) error
+}
+
+type Destination struct {
+	sdk.UnimplementedDestination
+	cfg          Config        // destination specific config for zendesk
+	buffer       []sdk.Record  // buffer stores the list of zendesk ticket from conduit server
+	ackFuncCache []sdk.AckFunc // returns error to conduit if fails else return nil
+	err          error         // to capture error at each method implementation
+	mux          *sync.Mutex   // maintains state of the pipeline
+	writer       Writer        // interface that implements to write tickets to zendesk
+}
+
+func NewDestination() sdk.Destination {
+	return &Destination{}
+}
+
+// Configure parses and initializes the config.
+func (d *Destination) Configure(ctx context.Context, cfg map[string]string) error {
+	configuration, err := Parse(cfg)
+	if err != nil {
+		return err
+	}
+
+	d.cfg = configuration
+	return nil
+}
+
+// Open http client
+func (d *Destination) Open(ctx context.Context) error {
+	d.mux = &sync.Mutex{}
+	d.buffer = make([]sdk.Record, 0, d.cfg.BufferSize)
+	d.ackFuncCache = make([]sdk.AckFunc, 0, d.cfg.BufferSize)
+	d.writer = zendesk.NewBulkImporter(d.cfg.UserName, d.cfg.APIToken, d.cfg.Domain, d.cfg.MaxRetries)
```

Comment:
Generally, it feels more intuitive to init. fields in a constructor or factory method (NewDestination() in this case).

Reply:
moved mutex to NewDestination.